core/src/main/java/io/questdb/cairo/wal/WalPurgeJob.java

```diff
@@ -583,4 +416,194 @@ protected boolean runSerially() {
         }
         return false;
     }
+
+    public interface Deleter {
+        void deleteSegmentDirectory(int walId, int segmentId);
+
+        void deleteWalDirectory(int walId);
+    }
+
+    public static class Logic {
+        private final StringSink debugBuffer = new StringSink();
+        private final Deleter deleter;
+        private final LongList discovered = new LongList();
+        private final IntIntHashMap nextToApply = new IntIntHashMap();
+        private final LongList nextToApplyKeys = new LongList();  // LongList rather than IntList because need .sort().
+        private TableToken tableToken;
+
+        public Logic(Deleter deleter) {
+            this.deleter = deleter;
+        }
+
+        public void accumDebugState() {
+            debugBuffer.clear();
```

Comment:
nit: you don't really need to save the state to buffer before execution run anymore since the state is immutable. The log message can be formatted instead

```suggestion
            LogRecord logBuffer = LOG.info();
            
            ....
            logBuffer.$();
            this.stateLogged = true;
```